### ISSUE TITLE
Performance Improvement on read/write cuts

### ISSUE THREAD

**Issue by Thuener** [odow/SDDP.jl#778]
When reading the cuts from file for a big model (with thousands of cuts) it can take couple of hours. My guess is because cut_selection is done every time we add a cut. Questions: 
- I was not able to find a way to disable cut_selection when reading the cuts. There is any way?
- We already did the cut selection when we first added the cuts; should we save only the cuts that we have already selected? This would reduce the cuts file size significantly and we would not have to do cut selection when reading the file.

**Comment by Thuener** [comment]
I can confirm that removing the cut_selection will reduce the time to load the model from 50 minutes to less than 1 minute.

**Comment by odow** [comment]
Adding a keyword to skip cut selection seems reasonable.

**Comment by Thuener** [comment]
I'm trying to refactor the read and write function without changing the other functions... 

My idea was to have something like this:
```julia
function write_cuts_to_file2(
    model::SDDP.PolicyGraph{T},
    filename::String;
    node_name_parser::Function = string,
    save_only_selected_cuts = false,
) where {T}
    @info "Using new write_cuts"
    cuts = Dict{String,Any}[]
    for (node_name, node) in model.nodes
        if node.objective_state !== nothing || node.belief_state !== nothing
            error(
                "Unable to write cuts to file because model contains " *
                "objective states or belief states.",
            )
        end
        node_cuts = Dict(
            "node" => node_name_parser(node_name),
            "single_cuts" => Dict{String,Any}[],
            "multi_cuts" => Dict{String,Any}[],
            "risk_set_cuts" => Vector{Float64}[],
        )
        oracle = node.bellman_function.global_theta
        for (cut, state) in zip(oracle.cuts, oracle.sampled_states)
            @info "$(node_name), cut.constraint_ref $(cut.constraint_ref)"
            if !save_only_selected_cuts || cut.constraint_ref !== nothing
                intercept = cut.intercept
                for (key, π) in cut.coefficients
                    intercept += π * state.state[key]
                end
                push!(
                    node_cuts["single_cuts"],
                    Dict(
                        "intercept" => intercept,
                        "coefficients" => copy(cut.coefficients),
                        "state" => copy(state.state),
                    ),
                )
            end
        end
        for (i, theta) in enumerate(node.bellman_function.local_thetas)
            for (cut, state) in zip(theta.cuts, theta.sampled_states)
                if !save_only_selected_cuts || cut.constraint_ref !== nothing
                    intercept = cut.intercept
                    for (key, π) in cut.coefficients
                        intercept += π * state.state[key]
                    end
                    push!(
                        node_cuts["multi_cuts"],
                        Dict(
                            "realization" => i,
                            "intercept" => intercept,
                            "coefficients" => copy(cut.coefficients),
                            "state" => copy(state.state),
                        ),
                    )
                end
            end
        end
        for p in node.bellman_function.risk_set_cuts
            push!(node_cuts["risk_set_cuts"], p)
        end
        push!(cuts, node_cuts)
    end
    open(filename, "w") do io
        return write(io, JSON.json(cuts))
    end
    return
end


function read_cuts_from_file2(
    model::SDDP.PolicyGraph{T},
    filename::String;
    node_name_parser::Function = SDDP._node_name_parser,
    use_cut_selection = true,
) where {T}
    @info "Using new read_cuts"
    cuts = JSON.parsefile(filename; use_mmap = false)
    for node_cuts in cuts
        node_name = node_name_parser(T, node_cuts["node"])::Union{Nothing,T}
        if node_name === nothing
            continue  # Skip reading these cuts
        end
        node = model[node_name]
        bf = node.bellman_function
        # Loop through and add the single-cuts.
        for json_cut in node_cuts["single_cuts"]
            has_state = haskey(json_cut, "state")
            state = if has_state
                Dict(Symbol(k) => v for (k, v) in json_cut["state"])
            else
                Dict(Symbol(k) => 0.0 for k in keys(json_cut["coefficients"]))
            end
            SDDP._add_cut(
                bf.global_theta,
                json_cut["intercept"],
                Dict(Symbol(k) => v for (k, v) in json_cut["coefficients"]),
                state,
                nothing,
                nothing;
                cut_selection = (use_cut_selection && has_state),
            )
        end
        
        # Loop through and add the multi-cuts. There are two parts:
        #  (i) the cuts w.r.t. the state variable x
        # (ii) the cuts that define the risk set
        # There is one additional complication: if these cuts are being read
        # into a new model, the local theta variables may not exist yet.
        if length(node_cuts["risk_set_cuts"]) > 0
            SDDP._add_locals_if_necessary(
                node,
                bf,
                length(first(node_cuts["risk_set_cuts"])),
            )
        end
        for json_cut in node_cuts["multi_cuts"]
            has_state = haskey(json_cut, "state")
            state = if has_state
                Dict(Symbol(k) => v for (k, v) in json_cut["state"])
            else
                Dict(Symbol(k) => 0.0 for k in keys(json_cut["coefficients"]))
            end
            bd_local_thetas_real = bf.local_thetas[json_cut["realization"]]
            SDDP._add_cut(
                bd_local_thetas_real,
                json_cut["intercept"],
                Dict(Symbol(k) => v for (k, v) in json_cut["coefficients"]),
                state,
                nothing,
                nothing;
                cut_selection = (use_cut_selection && has_state),
            )
        end
        
        # Here is part (ii): adding the constraints that define the risk-set
        # representation of the risk measure.
        for json_cut in node_cuts["risk_set_cuts"]
            expr = @expression(
                node.subproblem,
                bf.global_theta.theta - sum(
                    p * V.theta for (p, V) in zip(json_cut, bf.local_thetas)
                )
            )
            if JuMP.objective_sense(node.subproblem) == MOI.MIN_SENSE
                @constraint(node.subproblem, expr >= 0)
            else
                @constraint(node.subproblem, expr <= 0)
            end
        end
    end
    return
end
```
However, the code is a bit tangled. For example, the cuts are added to the ConvexApproximation just inside the cut selection part https://github.com/odow/SDDP.jl/blob/d2495d6a9a0886048d560df7e29c7b220656594c/src/plugins/bellman_functions.jl#L199
 Thus, without selection the model will not store the parameters of the cut, it will just add them to the model. This could lead to some issues, like not been able to write the cuts if you don't use cut selection.

**Comment by odow** [comment]
Let me take a look.

**Comment by odow** [comment]
So:

 * We should add a `cut_selection::Bool = true,` kwarg to `read_cuts_from_file`.

That might be enough to begin with. It doesn't solve the write them again issue.

 * We could add a `write_state::Bool = true,` kwarg to `write_cuts_to_file`. If you're not using the state, there's no need to write
 * We could add a `write_only_selected_cuts::Bool = false` to write only the cuts that are currently in the model.

I think I assumed people would want to always write the full set of cuts and then post-process them somehow after the fact.

**Comment by odow** [comment]
Closed by #781
